File: src/main/java/io/klbz/curie/SecondFocus.java
package io.klbz.curie;

import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;

public class SecondFocus<F, S> {
	private final Pair<F, S> pair;

	private SecondFocus(Pair<F, S> pair){ this.pair = pair; }

	public static <F, S> SecondFocus<F, S> on(Pair<F, S> pair){ return new SecondFocus<>(pair); }

	public Pair<F, S> unfocus(){ return pair; }

	public <SPrime> SecondFocus<F, SPrime> replace(SPrime newSecond){ return on(pair.replaceSecond(newSecond)); }

	public S isolate(){ return pair.isolateSecond(); }

	public <SPrime> SecondFocus<F, SPrime> map(Function<? super S, SPrime> f){ return on(pair.mapSecond(f)); }

	public SecondFocus<F, S> withDo(Consumer<S> doF){ return on(pair.withSecondDo(doF)); }

	@Override
	public boolean equals(Object o){
		if(this == o) return true;
		if(o == null) return false;

		if(o.getClass() == Pair.class) return Objects.equals(pair, o);

		if(o.getClass() == FirstFocus.class){
			FirstFocus<?, ?> that = (FirstFocus<?, ?>) o;
			return Objects.equals(pair, that.unfocus());
		}

		if(o.getClass() == SecondFocus.class){
			SecondFocus<?, ?> that = (SecondFocus<?, ?>) o;
			return Objects.equals(pair, that.pair);
		}

		return false;
	}

	@Override
	public int hashCode(){ return pair.hashCode(); }

	@Override
	public String toString(){ return "SecondFocus{" + pair + '}'; }
}

File: src/main/java/io/klbz/curie/SecondFocus.kt
package io.klbz.curie

import java.util.function.Consumer
import java.util.function.Function

class SecondFocus<F, S> private constructor(private val pair: Pair<F, S>) {

    fun unfocus() = pair

    fun <SPrime> replace(newSecond: SPrime) = on(pair.replaceSecond(newSecond))

    fun isolate() = pair.isolateSecond()

    fun <SPrime> map(f: Function<in S, SPrime>) = on(pair.mapSecond(f))

    fun withDo(doF: Consumer<S>) = on(pair.withSecondDo(doF))

    //region Kotlin-specific overloads

    @JvmSynthetic
    fun <SPrime> map(f: (S) -> SPrime) = on(pair.mapSecond(f))

    @JvmSynthetic
    fun withDo(doF: (S) -> kotlin.Unit) = also { pair.withSecondDo(doF) }

    //endregion

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false

        if (other.javaClass == Pair::class.java) return pair == other

        if (other.javaClass == FirstFocus::class.java) {
            val that = other as FirstFocus<*, *>?
            return pair == that!!.unfocus()
        }

        if (other.javaClass == SecondFocus::class.java) {
            val that = other as SecondFocus<*, *>?
            return pair == that!!.pair
        }

        return false
    }

    override fun hashCode() = pair.hashCode()

    override fun toString() = "SecondFocus{$pair}"

    companion object {
        @JvmStatic
        fun <F, S> on(pair: Pair<F, S>) = SecondFocus(pair)
    }
}
